### PR TITLE
[CDAP-16038] optimize graphql queries

### PR DESCRIPTION
**Pull request by elfenheart**
JIRAs:
- https://issues.cask.co/browse/CDAP-16038
- https://issues.cask.co/browse/CDAP-16074

Build: https://builds.cask.co/browse/CDAP-URUT196

### CODE REVIEW

File: cdap-ui/package.json [hunk 1]
```diff
@@ -161,6 +162,7 @@
     "cookie-parser": "1.4.1",
     "css-vars-ponyfill": "1.8.0",
     "d3": "4.12.2",
+    "dataloader": "^1.4.0",
```
Comment (ajainarayanan) [inline cdap-ui/package.json]: nit: This should be single specific version

File: cdap-ui/graphql/helpers/BatchEndpoints/programRuns.js [hunk 2]
```diff
@@ -18,34 +18,28 @@ const urlHelper = require('../../../server/url-helper'),
...
+  options.body = req.map((reqObj) => reqObj.program);
 
-  if (runInfo.length === 0) {
-    return [];
-  }
+  const runInfo = await resolversCommon.requestPromiseWrapper(options, auth);
 
-  return runInfo[0].runs;
+  return options.body.map((program) => {
+    return find(runInfo, (run) => {
```
Comment (ajainarayanan) [inline cdap-ui/graphql/helpers/BatchEndpoints/programRuns.js]: Won't this be efficient if we only request runs for workflow and spark programs? We know the workflow and spark programs are constant(ish). This is kind of N^2 operation here right?

**Review by ajainarayanan — APPROVED** [review]
Minor comment. Otherwise 👍

File: cdap-ui/graphql/helpers/BatchEndpoints/programRuns.js (hunk 2 above)
Comment (elfenheart) [reply cdap-ui/graphql/helpers/BatchEndpoints/programRuns.js]: We are only requesting the runs for workflow and spark depending on the pipeline type. So essentially what's happening with this, the DataLoader requires the request array and the response array to be in the same order. However, the backend do not actually respond with the same order.

This was a quick way to solve it, but yes it's not efficient. We can generate a map of pipeline name to the info, then have this return map to just do a lookup from that map, then it's O(n)

**Review by ajainarayanan — APPROVED** [review]

**Pull request merged**